File: sdk/java/src/com/ouc/tcp/test/TestSender.java
package com.ouc.tcp.test;

import com.ouc.tcp.sdk.*;
import java.util.Arrays;

public class TestSender extends TransportProtocol {

    @Override
    public void init(SystemContext ctx) {
        ctx.log("Java TestSender initialized!");
    }

    @Override
    public void onPacket(SystemContext ctx, Packet packet) {
        ctx.log("Java TestSender received ACK: " + packet.getHeader().getAckNum());
    }

    @Override
    public void onTimer(SystemContext ctx, int timerId) {
        ctx.log("Java TestSender timer expired: " + timerId);
    }

    @Override
    public void onAppData(SystemContext ctx, byte[] data) {
        ctx.log("Java TestSender sending data: " + Arrays.toString(data));
        
        TcpHeader h = new TcpHeader();
        h.setSeqNum(1);
        h.setAckNum(0);
        h.setFlags((byte)0);
        h.setWindowSize(100);
        
        Packet p = new Packet(h, data);
        ctx.sendPacket(p);
        
        // Start a timer just for fun
        ctx.startTimer(1000, 1);
    }
}

File: sdk/cpp/examples/TestSender.cpp
// Simple example C++ sender using the SDK.
// This file demonstrates how to implement a protocol in C++ and expose
// the standard entrypoints expected by the Rust loader.

#include <ouc/tcp/sdk/TransportProtocol.hpp>
#include <ouc/tcp/sdk/EntryPoints.hpp>

using namespace ouc::tcp::sdk;

class TestSender : public TransportProtocol {
public:
    void init(SystemContext& ctx) override {
        ctx.log("C++ TestSender init");
    }

    void onAppData(SystemContext& ctx, const std::vector<std::uint8_t>& data) override {
        TcpHeader h;
        h.seqNum = nextSeq_;
        Packet p{h, data};
        ctx.log("C++ TestSender sending packet");
        ctx.sendPacket(p);
        ++nextSeq_;
    }

    void onPacket(SystemContext& ctx, const Packet& packet) override {
        (void)ctx;
        (void)packet;
        // For a simple test sender we ignore incoming packets.
    }

    void onTimer(SystemContext& ctx, int timerId) override {
        (void)ctx;
        (void)timerId;
    }

private:
    std::uint32_t nextSeq_{0};
};

// Generate create_sender/destroy_sender and the sender_* C entrypoints
// that the Rust loader expects.
TCP_LAB_DEFINE_PROTOCOL_ENTRYPOINTS(TestSender);

